Add Client Side Certificate Support

Lokesh-K-Haralakatta opened the pull request:
The code changes include client side certificate support for device and gateway. Tested the changes with the supported underlying Node JS library, worked fine.

knolleary commented:
Hi,

Rather than recreate your own tls node and duplicate what node-red already provides, you should just use the existing `tls-config` node.

That way your node will benefit from the improvements to the core `tls-config` node as we continue to enhance it.

Nick

Lokesh-K-Haralakatta commented:
Hi Nick,
First, I looked to reuse the tls-config node but as it uses the MQTT library to set the TLS options, we can not directly reuse that. We need to tweak to use our Node JS client library to make use of underlying Client Side Certificate support present in Node JS library. So, we need to have our own TLS node with required changes.

Lokesh-K-Haralakatta commented:
Hi Nick,

There are 2 things in the tls-config node's code,

Function addTLSOptions, this links to MQTT library

Second, most important one is that we don't want to read the certificate contents in the tls node itself because, we have the logic of reading the  certificates contents in our underlying Node JS library. We just need paths to certificates.

These are 2 points, which will not allow us to reuse the tls-config node as is.

I have already added the Server CA entry and made that as optional.

if(tlsNode.config.serverPath !== ''){
                   this.config['server-ca'] = tlsNode.config.serverPath;
                }

Regd the name - tls-config, I can change that to wiotp-tls-config.

knolleary commented:
Sorry, where exactly does `addTLSOptions` go anywhere near the MQTT library? It is a utility function that, given an object, sets the standard node.js tls options that correspond to the different values provided. No where does the node load the certificate contents in - it passes the paths to the files back.

Lokesh-K-Haralakatta commented:
addTLSOptions is utility function and doesn't link to MQTT library, that's fine. The second point I"m referring to this code in 05-tls.js:

try {
            if (certPath) {
                this.cert = fs.readFileSync(certPath);
            }
            if (keyPath) {
                this.key = fs.readFileSync(keyPath);
            }
            if (caPath) {
                this.ca = fs.readFileSync(caPath);
            }
        } catch(err) {
            this.valid = false;
            this.error(err.toString());
            return;
        }

This code is not required for us in tls node, as we already have in Node JS library.

knolleary commented:
Ah right, sorry. I see.

The standard node.js tls object supports either file paths or actual certificate data. I would suggest your library ought to match as that is what users would expect.

Lokesh-K-Haralakatta commented:
For end users, there is no difference. They just need to  provide the paths to certificates even with Node JS library. Since we are using Node JS as underlying library to connect to WIoTP, we need to have the logic to read certificates path in Node JS library because sometimes users may use only Node JS library directly to connect to WIoTP. Hence, we can not have reading logic in TLS Node. If you agree with this, then will go ahead and rename the node to wiotp-tls-config and submit for merge?

knolleary commented:
I don't understand. If you are using node.js, which supports both paths and loaded certificates, why do you need the logic to read the certs yourself?

knolleary commented:
Forgive me, just double checked and I see the node.js lib does not support paths.
---
In our next release, we'll be adding support for the user to upload their certs to the node directly, rather than provide a file path. This is important in Bluemix because forcing the user to add files to their app is a big hurdle.

All I can say, is I strongly encourage you to reuse the core node and not duplicate effort. That may mean and update to your library to handle paths and certs, but it would, in my opinion, be worth it.

Not much else I can add.

Lokesh-K-Haralakatta commented:
Hi Nick,
I don't think any library supports reading both Paths to Certificates and Certificates Contents. I have seen with mqtt.js, Python client from Paho, mbedTLS, all these libraries just take path to certificates not the certificates contents from the end user. I don't think, it's good idea to do this.

The effort needed to support both for Node.js is much more than putting effort to create one more node for TLS Config. So, what I'm thinking is will try to get the paths to certificates from the existing TLS node and feed to Node.js library. That way, we meet everyone's arguments.

Lokesh-K-Haralakatta commented:
Hi Nick,
I have added required changes to Node.js library to support both certificates path and certificates contents (this is internal, not exposed to end user). With this change, I'm able to reuse the tls-config node in our Watson IoT Node. I have added the required changes and tested, worked fine. Please review the changes and let me know if you have any comments?
